File: app/src/main/java/com/vitaliyhtc/dagger2investigation/domain/ProductRepository.java
package com.vitaliyhtc.dagger2investigation.domain;

import com.vitaliyhtc.dagger2investigation.domain.model.Product;

import java.util.List;

import io.reactivex.Single;

public interface ProductRepository {

    Single<List<Product>> getProducts(int page);

    Single<Product> getOneProduct(int productId);
}

File: app/src/main/java/com/vitaliyhtc/dagger2investigation/domain/ProductRepository.kt
package com.vitaliyhtc.dagger2investigation.domain

import android.arch.lifecycle.LifecycleOwner
import android.support.test.espresso.idling.CountingIdlingResource
import com.vitaliyhtc.dagger2investigation.domain.model.Product
import io.reactivex.Single

interface ProductRepository {

    fun setIdlingResource(idlingResource: CountingIdlingResource)

    fun getProducts(page: Int): Single<List<Product>>

    fun subscribeForProductsUpdates(lifecycleOwner: LifecycleOwner, listener: (products: List<Product>) -> Unit)

    fun subscribeForProductsUpdates(listener: (products: List<Product>) -> Unit)

    fun getOneProduct(productId: Int): Single<Product>

    fun updateProduct(product: Product)
}